iterative_refinement/expectation_maximization.py

```diff
@@ -0,0 +1,435 @@
+
+                # return inserted_slice_3d, count_3d
+                # TODO: remove placeholder
+                return slice_real, xyz, n_pix
+
+            for one_slice_idx in range(bayes_factors_1.shape[0]):
+                xyz = xyz_rotated[one_slice_idx]
+
+                # if this can be vectorized, can avoid loop over slices
+                inserted_slice_3d_r, count_3d_r = do_insert_slice(
+                    particle_f_deconv_1.real, xyz, np.zeros((n_pix, n_pix, n_pix))
+                )
+                inserted_slice_3d_i, count_3d_i = do_insert_slice(
+                    particle_f_deconv_1.imag, xyz, np.zeros((n_pix, n_pix, n_pix))
+                )
+
+                imag = 1j * inserted_slice_3d_i
+                map_3d_f_updated_1 += inserted_slice_3d_r + imag
+                counts_3d_updated_1 += count_3d_r + count_3d_i
+
+            for one_slice_idx in range(bayes_factors_2.shape[0]):
+                xyz = xyz_rotated[one_slice_idx]
+
+                # if this can be vectorized, can avoid loop over slices
+                inserted_slice_3d_r, count_3d_r = do_insert_slice(
+                    particle_f_deconv_2.real, xyz, np.zeros((n_pix, n_pix, n_pix))
+                )
+                inserted_slice_3d_i, count_3d_i = do_insert_slice(
+                    particle_f_deconv_2.imag, xyz, np.zeros((n_pix, n_pix, n_pix))
+                )
+                imag = 1j * inserted_slice_3d_i
+                map_3d_f_updated_2 += inserted_slice_3d_r + imag
+                counts_3d_updated_2 += count_3d_r + count_3d_i
+
+        # apply noise model
+        # half_map_1, half_map_2 come from doing the above independently
+        # filter by noise estimate (e.g. multiply both half maps by FSC)
+
+        def do_fsc(map_3d_f_1, map_3d_f_2):
+            """
+            Compute Fourier shell correlation.
+            Estimate noise from half maps.
+
+            Parameters
+            ----------
+            map_3d_f_1 : arr
+                Shape (n_pix, n_pix, n_pix)
+            map_3d_f_2 : arr
+                Shape (n_pix, n_pix, n_pix)
+
+            Returns
+            -------
+            fsc_1d_1 : arr
+                Noise estimates for map 1.
+                Shape (n_pix // 2,)
+            fsc_1d_2 : arr
+                Noise estimates for map 2.
+                Shape (n_pix // 2,)
+            """
+            # TODO: write fast vectorized fsc from code snippets in
+            # https://github.com/geoffwoollard/learn_cryoem_math/blob/master/nb/fsc.ipynb
+            # https://github.com/geoffwoollard/learn_cryoem_math/blob/master/nb/mFSC.ipynb
+            # https://github.com/geoffwoollard/learn_cryoem_math/blob/master/nb/guinier_fsc_sharpen.ipynb
+            n_pix_1 = map_3d_f_1.shape[0]
+            n_pix_2 = map_3d_f_2.shape[0]
+            fsc_1d_1 = np.ones(n_pix_1 // 2)
+            fsc_1d_2 = np.ones(n_pix_2 // 2)
+            return fsc_1d_1, fsc_1d_2
```

Comment:
There will only be one correlation, since it's a function of both `map_3d_f_1` and `map_3d_f_2`. 

`n_pix_1` is guaranteed to be `n_pix_2` by what happens upstream.

```
n_pix = map_3d_f_1.shape[0]
fsc_1d = np.ones(n_pix // 2)
return fsc_1d
```